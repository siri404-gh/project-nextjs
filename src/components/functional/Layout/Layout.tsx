import React, { useState, Fragment, FunctionComponent } from 'react';
import { makeStyles } from '@material-ui/core';
import { useRouter } from 'next/router';

import Navbar, { Props as navbarPropsType } from '@/components/functional/Navbar/Navbar';
import BottomBar, { Props as bottomBarPropsType, bottomListType } from '@/components/functional/Bottombar/Bottombar';
import Sidebar, { Props as sidebarPropsType } from '@/components/functional/Sidebar/Sidebar';
import SidebarContent, { Props as sidebarContentPropsType } from '@/components/functional/SidebarContent/SidebarContent';
import Seo, { Props as seoPropsType } from '@/components/functional/Seo/Seo';

export interface Props {
  seoProps?: seoPropsType,
  navbarProps?: navbarPropsType,
  sidebarProps?: sidebarPropsType,
  sidebarContentProps?: sidebarContentPropsType,
  bottomBarProps?: bottomBarPropsType,
  children?: JSX.Element,
  index?: number,
  sections?: bottomListType[],
}

const useStyles = makeStyles(() => ({
  bottomNav: {
    position: 'fixed',
    bottom: 0,
    left: 0,
    width: '100%',
    borderTop: 'solid 1px #e4e4e4',
  },
  contentWrapper: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    display: 'flex',
    flexDirection: 'column',
    width: '100%',
  },
  shift: {
    minHeight: 56,
    maxHeight: 56,
  },
}));

const Layout: FunctionComponent<Props> = ({ seoProps, navbarProps, sidebarProps, sidebarContentProps, bottomBarProps, children, sections }) => {
  const classes = useStyles();
  const router = useRouter();
  const [ isSidebarOpen, setIsSidebarOpen ] = useState(false);

  const onNavbarLeftButtonClick = () => setIsSidebarOpen(true);
  const onNavbarTitlebarClick = () => router.push('/');
  const onSidebarOpen = () => setIsSidebarOpen(true);
  const onSidebarClose = () => setIsSidebarOpen(false);
  const onBottomNavItemChange = val => router.push(sections[val].route);

  return <Fragment>
    <Seo {...seoProps} />
    <Navbar
      {...navbarProps}
      onLeftButtonClick={onNavbarLeftButtonClick}
      onTitlebarClick={onNavbarTitlebarClick} />
    <Sidebar
      {...sidebarProps}
      isOpen={isSidebarOpen}
      onClose={onSidebarClose}
      onOpen={onSidebarOpen}>
      <SidebarContent
        {...sidebarContentProps} />
    </Sidebar>
    <div className={classes.contentWrapper}>
      <div className={classes.shift} />
      {children}
      <div className={classes.shift} />
    </div>
    <BottomBar
      {...bottomBarProps}
      onChange={onBottomNavItemChange}
      className={classes.bottomNav} />
  </Fragment>
};

Layout.defaultProps = {
  children: null,
  index: 0,
  sections: [],
};

export default Layout;
