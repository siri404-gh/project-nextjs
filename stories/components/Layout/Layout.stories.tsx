import React from 'react';
import { Story } from '@storybook/react/types-6-0';

import Layout, { Props } from '@/components/functional/Layout/Layout';

export default {
  title: 'Layout',
  component: Layout,
};

const Template: Story<Props> = args => <Layout {...args} />;

export const FirstStory = Template.bind({});

FirstStory.args = {
  sections: [
    { 'title': '1', 'route': '/route', 'img': 'https://fakeimg.pl/300/' },
    { 'title': '2', 'route': '/route', 'img': 'https://fakeimg.pl/300/' },
    { 'title': '3', 'route': '/route', 'img': 'https://fakeimg.pl/300/' },
    { 'title': '4', 'route': '/route', 'img': 'https://fakeimg.pl/300/' },
  ],
  links: [
    { 'title': 'Title', 'route': '/route', 'label': 'label' }
  ]
};
