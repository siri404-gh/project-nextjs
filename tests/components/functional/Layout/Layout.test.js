import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { useRouter } from 'next/router';

import Layout from '@/components/functional/Layout/Layout';

jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}))

const mockRouter = {
  push: jest.fn(),
};

useRouter.mockReturnValue(mockRouter);

describe('Layout', () => {
  let container;
  const sections = [{
    title: 'title',
    route: '/route',
    content: 'content',
  }];

  let props = {
    sections,
    bottomBarProps: {
      list: sections,
    },
  };

  beforeEach(() => {
    container = document.createElement('section');
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('Should render without props', () => {
    act(() => {
      render(<Layout />, container);
    });
    expect(container).toMatchSnapshot();
  });

  it('Should redirect to home when titlebar is clicked', () => {
    act(() => {
      render(<Layout />, container);
      const titlebar = document.querySelector(".Navbar__titlebar");
      // titleBar.dispatchEvent(new MouseEvent("click"), { bubbles: true });
      titlebar.click();
    });
    expect(mockRouter.push).toHaveBeenCalledWith('/');
  });

  it('Should call onBottomNavItemChange when button in the bottom nav is clicked', () => {
    act(() => {
      render(<Layout {...props} />, container);
      const button = Array.from(document.querySelectorAll(".MuiBottomNavigationAction-root"))[0];
      button.click();
    });
    expect(mockRouter.push).toHaveBeenCalledWith('/route');
  });

  it('Should open sidebar when left button in navbar is clicked', () => {
    act(() => {
      render(<Layout />, container);
      const button = document.querySelector(".Navbar__leftbutton");
      button.click();
    });
    expect(container).toMatchSnapshot();
  });

  it('Should close sidebar when clicked outside sidebar', () => {
    act(() => {
      render(<Layout />, container);
      const button = document.querySelector(".Navbar__leftbutton");
      button.click();
    });
    act(() => {
      document.querySelector('.MuiBackdrop-root').click();
    });
    expect(container).toMatchSnapshot();
  });

});
