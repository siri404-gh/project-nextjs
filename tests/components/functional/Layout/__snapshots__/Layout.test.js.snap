// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Should close sidebar when clicked outside sidebar 1`] = `
<section>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-94 mui-fixed MuiPaper-elevation0"
    style=""
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-100 MuiToolbar-gutters"
    >
      <button
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root Navbar__leftbutton makeStyles-leftBtn-95"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="Navbar__titlebar makeStyles-titlebar-97"
      >
        <div
          class="makeStyles-titleLeft-96"
        />
        <button
          aria-controls="simple-menu"
          aria-hidden="true"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-small-101 MuiSvgIcon-colorSecondary"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
  <div
    class="PrivateSwipeArea-root-107 PrivateSwipeArea-anchorLeft-108"
    style="width: 50px;"
  />
  <div
    class="makeStyles-contentWrapper-92"
  >
    <div
      class="makeStyles-shift-93"
    />
    <div
      class="makeStyles-shift-93"
    />
  </div>
</section>
`;

exports[`Layout Should open sidebar when left button in navbar is clicked 1`] = `
<section
  aria-hidden="true"
>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-67 mui-fixed MuiPaper-elevation0"
    style="padding-right: 0px;"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-73 MuiToolbar-gutters"
    >
      <button
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root Navbar__leftbutton makeStyles-leftBtn-68"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="Navbar__titlebar makeStyles-titlebar-70"
      >
        <div
          class="makeStyles-titleLeft-69"
        />
        <button
          aria-controls="simple-menu"
          aria-hidden="true"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-small-74 MuiSvgIcon-colorSecondary"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
  <div
    class="PrivateSwipeArea-root-80 PrivateSwipeArea-anchorLeft-81"
    style="width: 50px;"
  />
  <div
    class="makeStyles-contentWrapper-65"
  >
    <div
      class="makeStyles-shift-66"
    />
    <div
      class="makeStyles-shift-66"
    />
  </div>
</section>
`;

exports[`Layout Should render without props 1`] = `
<section>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionFixed MuiAppBar-colorPrimary makeStyles-appBar-4 mui-fixed MuiPaper-elevation0"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-10 MuiToolbar-gutters"
    >
      <button
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root Navbar__leftbutton makeStyles-leftBtn-5"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <div
        class="Navbar__titlebar makeStyles-titlebar-7"
      >
        <div
          class="makeStyles-titleLeft-6"
        />
        <button
          aria-controls="simple-menu"
          aria-hidden="true"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root makeStyles-small-11 MuiSvgIcon-colorSecondary"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </header>
  <div
    class="PrivateSwipeArea-root-17 PrivateSwipeArea-anchorLeft-18"
    style="width: 50px;"
  />
  <div
    class="makeStyles-contentWrapper-2"
  >
    <div
      class="makeStyles-shift-3"
    />
    <div
      class="makeStyles-shift-3"
    />
  </div>
</section>
`;
